utils/cScreenInit.cpp

```diff
@@ -0,0 +1,173 @@
+#endif
+
+// tries them in this order (1st come first serve)
+// do note, when DPI is > 100% in OS (atleast in Windows) things seem to get wonky.
+// see also: https://github.com/stefanhendriks/Dune-II---The-Maker/issues/314
+std::array<ScreenResolution, 9> kResolutionsToTry = {
+    ScreenResolution{3840, 2160}, // 4k
+    ScreenResolution{2048, 1536}, // 4:3 aspect ratio
+    ScreenResolution{1920, 1080}, // Full HD
+    ScreenResolution{1920, 1440}, // 4:3 aspect ratio
+    ScreenResolution{1600, 900},
+    ScreenResolution{1280, 1024},
+    ScreenResolution{1366, 768},  // widescreen
+    ScreenResolution{1024, 768},
+    ScreenResolution{800, 600}
+};
+
+const GFX_MODE* FindGfxMode(const GFX_MODE_LIST& list, ScreenResolution res, int colorDepth) {
+    for (int n = 0; n < list.num_modes; ++n) {
+        const GFX_MODE& mode = list.mode[n];
+        if (mode.width == res.width && mode.height == res.height && mode.bpp == colorDepth) {
+            return &list.mode[n];
+        }
+    }
+    return nullptr;
+}
+
+template <std::size_t N>
+std::optional<ScreenResolution> TryAndSetFirstValidResolution(std::span<ScreenResolution, N> resolutionsToTry) {
+    auto logger = cLogger::getInstance();
+
+    for (ScreenResolution res : resolutionsToTry) {
+        const auto msg = fmt::format("Trying to init full-screen at {}x{}.", res.width, res.height);
+        logger->log(LOG_INFO, COMP_ALLEGRO, "Screen init", msg);
+
+        const int r = set_gfx_mode(GFX_AUTODETECT, res.width, res.height, res.width, res.height);
+
+        if (r > -1) {
+            return res;
+        }
+    }
+    return {};
+}
+
+std::optional<ScreenResolution> SetBestScreenResolution(int colorDepth) {
+    // Allegro supports the following color depths
+    assert(colorDepth == 8 || colorDepth == 15 || colorDepth == 16 ||
+           colorDepth == 24 || colorDepth == 32);
+
+    auto logger = cLogger::getInstance();
+
+    // Note: do not use "magic" drivers like GFX_AUTODETECT here, that doesn't work according to the Allegro
+    // documentation.
+    GFX_MODE_LIST* modeList = get_gfx_mode_list(kNonMagicVideoDriver);
+
+    std::optional<ScreenResolution> setResolution;
+    if (modeList) {
+        // Make a shortlist of resolutions to try by filtering unsupported ones.
+        std::vector<ScreenResolution> supportedResolutions;
+        for (ScreenResolution res : kResolutionsToTry) {
+            auto mode = FindGfxMode(*modeList, res, colorDepth);
+            if (mode) {
+                supportedResolutions.emplace_back(res);
+            }
+        }
+
+        destroy_gfx_mode_list(modeList);
+
+        return TryAndSetFirstValidResolution(std::span(supportedResolutions));
+    } else {
+        logger->log(LOG_WARN, COMP_ALLEGRO, "Screen auto init",
+                    "Allegro could not determine suppored resolutions. Trying out several of them.");
+        return TryAndSetFirstValidResolution(std::span(kResolutionsToTry));
+    }
+}
+
+}
```

Comment:
or wait, does that namespace end here?

Reply:
Yes, it's an anonymous namespace. Stuff in anonymous namespaces has local linkage. This means it can't be accessed from outside of the translation unit (.cpp file in this case). Saves the linker some effort.